ansible_collections/f5networks/f5_modules/plugins/modules/bigip_device_info.py

```diff
@@ -7553,6 +7556,57 @@ def read_collection_from_device(self):
         return result
 
 
+class As3Parameters(BaseParameters):
+    api_map = {
+    }
+
+    returnables = [
+
+    ]
+
+
+class As3FactManager(BaseManager):
+    def __init__(self, *args, **kwargs):
+        self.client = kwargs.get('client', None)
+        self.module = kwargs.get('module', None)
+        super(As3FactManager, self).__init__(**kwargs)
+        self.want = self.module.params
+
+    def exec_module(self):
+        facts = self._exec_module()
+        result = dict(as3_config=facts)
+        return result
+
+    def _exec_module(self):
+        results = []
+        facts = self.read_facts()
+        return facts
+
+    def read_facts(self):
+        results = []
+        collection = self.read_collection_from_device()
+        return collection
+
+    def read_collection_from_device(self):
+        uri = "https://{0}:{1}/mgmt/shared/appsvcs/declare".format(
+            self.client.provider['server'],
+            self.client.provider['server_port'],
+        )
+        resp = self.client.api.get(uri)
+        try:
+            response = resp.json()
+        except ValueError as ex:
+            raise F5ModuleError(str(ex))
+
+        if resp.status not in [200, 201] or 'code' in response and response['code'] not in [200, 201]:
+            raise F5ModuleError(resp.content)
+
+        if 'class' not in response:
+            return []
+        result = response
+        return result
```

Comment:
last time i checked the entire declaration sits under declaration? we are only interested in that nothing else

Reply:
hmmm no, it doesn't... It does this weird thing where the tenants are separated out like this - 

```
{
  "Sample_01": {
    "class": "Tenant",
    "A1": {
      "class": "Application",
      "template": "http",
      "serviceMain": {
        "class": "Service_HTTP",
        "virtualAddresses": [
          "10.0.1.10"
        ],
        "pool": "web_pool"
      },
      "web_pool": {
        "class": "Pool",
        "monitors": [
          "http"
        ],
        "members": [
          {
            "servicePort": 80,
            "serverAddresses": [
              "192.0.1.10",
              "192.0.1.11"
            ]
          }
        ]
      }
    }
  },
  "class": "ADC",
  "schemaVersion": "3.0.0",
  "id": "urn:uuid:33045210-3ab8-4636-9b2a-c98d22ab915d",
  "label": "Sample 1",
  "remark": "Simple HTTP Service with Round-Robin Load Balancing",
  "updateMode": "selective",
  "controls": {
    "archiveTimestamp": "2020-05-23T02:33:47.506Z"
  }
}
```

Reply:
A better example to show you what i mean, here is the device output:
```
{
    "class": "ADC",
    "schemaVersion": "3.0.0",
    "id": "urn:uuid:33045210-3ab8-4636-9b2a-c98d22ab915d",
    "label": "Sample 1",
    "remark": "Simple HTTP application with RR pool",
    "Sample_01": {
        "class": "Tenant",
        "A1": {
            "class": "Application",
            "template": "http",
            "serviceMain": {
                "class": "Service_HTTP",
                "virtualAddresses": [
                    "10.0.1.10"
                ],
                "pool": "web_pool",
                "virtualPort": 80,
                "persistenceMethods": [
                    "cookie"
                ],
                "profileHTTP": "basic",
                "virtualType": "standard",
                "layer4": "tcp",
                "profileTCP": "normal",
                "serviceDownImmediateAction": "none",
                "shareAddresses": false,
                "enable": true,
                "maxConnections": 0,
                "snat": "auto",
                "addressStatus": true,
                "mirroring": "none",
                "lastHop": "default",
                "translateClientPort": false,
                "translateServerAddress": true,
                "translateServerPort": true,
                "nat64Enabled": false
            },
            "web_pool": {
                "class": "Pool",
                "monitors": [
                    "http"
                ],
                "members": [
                    {
                        "servicePort": 80,
                        "serverAddresses": [
                            "192.0.1.10",
                            "192.0.1.11"
                        ],
                        "enable": true,
                        "connectionLimit": 0,
                        "rateLimit": -1,
                        "dynamicRatio": 1,
                        "ratio": 1,
                        "priorityGroup": 0,
                        "adminState": "enable",
                        "addressDiscovery": "static",
                        "shareNodes": false
                    }
                ],
                "loadBalancingMode": "round-robin",
                "minimumMembersActive": 1,
                "reselectTries": 0,
                "serviceDownAction": "none",
                "slowRampTime": 10,
                "minimumMonitors": 1
            },
            "enable": true
        },
        "enable": true,
        "defaultRouteDomain": 0,
        "optimisticLockKey": ""
    },
    "Sample_02": {
        "class": "Tenant",
        "A1": {
            "class": "Application",
            "template": "http",
            "serviceMain": {
                "class": "Service_HTTP",
                "virtualAddresses": [
                    "10.0.1.11"
                ],
                "pool": "web_pool2",
                "virtualPort": 80,
                "persistenceMethods": [
                    "cookie"
                ],
                "profileHTTP": "basic",
                "virtualType": "standard",
                "layer4": "tcp",
                "profileTCP": "normal",
                "serviceDownImmediateAction": "none",
                "shareAddresses": false,
                "enable": true,
                "maxConnections": 0,
                "snat": "auto",
                "addressStatus": true,
                "mirroring": "none",
                "lastHop": "default",
                "translateClientPort": false,
                "translateServerAddress": true,
                "translateServerPort": true,
                "nat64Enabled": false
            },
            "web_pool2": {
                "class": "Pool",
                "monitors": [
                    "http"
                ],
                "members": [
                    {
                        "servicePort": 80,
                        "serverAddresses": [
                            "192.0.1.12",
                            "192.0.1.13"
                        ],
                        "enable": true,
                        "connectionLimit": 0,
                        "rateLimit": -1,
                        "dynamicRatio": 1,
                        "ratio": 1,
                        "priorityGroup": 0,
                        "adminState": "enable",
                        "addressDiscovery": "static",
                        "shareNodes": false
                    }
                ],
                "loadBalancingMode": "round-robin",
                "minimumMembersActive": 1,
                "reselectTries": 0,
                "serviceDownAction": "none",
                "slowRampTime": 10,
                "minimumMonitors": 1
            },
            "enable": true
        },
        "enable": true,
        "defaultRouteDomain": 0,
        "optimisticLockKey": ""
    },
    "updateMode": "selective",
    "controls": {
        "archiveTimestamp": "2020-08-18T10:05:46.156Z",
        "class": "Controls",
        "logLevel": "error",
        "trace": false,
        "traceResponse": false
    }
}
```
Here is what needs to be found in the output from bigip_device_info, obviously we are only interested what is under declaration key, in future i plan to slice that output to split by tenant but we might not do this with the module but write a filter plugin for this
```
{
  "declaration": {
    "class": "ADC",
    "schemaVersion": "3.0.0",
    "id": "urn:uuid:33045210-3ab8-4636-9b2a-c98d22ab915d",
    "label": "Sample 1",
    "remark": "Simple HTTP application with RR pool",
    "Sample_01": {
      "class": "Tenant",
      "A1": {
        "class": "Application",
        "template": "http",
        "serviceMain": {
          "class": "Service_HTTP",
          "virtualAddresses": [
            "10.0.1.10"
          ],
          "pool": "web_pool"
        },
        "web_pool": {
          "class": "Pool",
          "monitors": [
            "http"
          ],
          "members": [{
            "servicePort": 80,
            "serverAddresses": [
              "192.0.1.10",
              "192.0.1.11"
            ]
          }]
        }
      }
    },
    "Sample_02": {
      "class": "Tenant",
      "A1": {
        "class": "Application",
        "template": "http",
        "serviceMain": {
          "class": "Service_HTTP",
          "virtualAddresses": [
            "10.0.1.11"
          ],
          "pool": "web_pool2"
        },
        "web_pool2": {
          "class": "Pool",
          "monitors": [
            "http"
          ],
          "members": [{
            "servicePort": 80,
            "serverAddresses": [
              "192.0.1.12",
              "192.0.1.13"
            ]
          }]
        }
      }
    }
  }
}
```

Reply:
If all you're asking me to do is wrap it in a declaration key, I wrap it in a key called as3_config/do_config/ts_config right now. I can rename it to "declaration". I thought you meant I need to remove the class, schemaVersion etc, which is just an implementation of the _remove_internal_keywords function, and return only the parts in Sample_01, Sample_02. Since that's not the case, this is what my current output looks like - 

```
TASK [bigip_device_info : debug] **********************************************************************************************
ok: [bigip_15] => {
    "result": {
        "ansible_facts": {
            "ansible_net_as3_config": {
                "Sample_01": {
                    "A1": {
                        "class": "Application",
                        "serviceMain": {
                            "class": "Service_HTTP",
                            "pool": "web_pool",
                            "virtualAddresses": [
                                "10.0.1.10"
                            ]
                        },
                        "template": "http",
                        "web_pool": {
                            "class": "Pool",
                            "members": [
                                {
                                    "serverAddresses": [
                                        "192.0.1.10",
                                        "192.0.1.11"
                                    ],
                                    "servicePort": 80
                                }
                            ],
                            "monitors": [
                                "http"
                            ]
                        }
                    },
                    "class": "Tenant"
                },
                "class": "ADC",
                "controls": {
                    "archiveTimestamp": "2020-05-23T02:33:47.506Z"
                },
                "id": "urn:uuid:33045210-3ab8-4636-9b2a-c98d22ab915d",
                "label": "Sample 1",
                "remark": "Simple HTTP Service with Round-Robin Load Balancing",
                "schemaVersion": "3.0.0",
                "updateMode": "selective"
            },
            "ansible_net_queried": true
        },
        "as3_config": {
            "Sample_01": {
                "A1": {
                    "class": "Application",
                    "serviceMain": {
                        "class": "Service_HTTP",
                        "pool": "web_pool",
                        "virtualAddresses": [
                            "10.0.1.10"
                        ]
                    },
                    "template": "http",
                    "web_pool": {
                        "class": "Pool",
                        "members": [
                            {
                                "serverAddresses": [
                                    "192.0.1.10",
                                    "192.0.1.11"
                                ],
                                "servicePort": 80
                            }
                        ],
                        "monitors": [
                            "http"
                        ]
                    }
                },
                "class": "Tenant"
            },
            "class": "ADC",
            "controls": {
                "archiveTimestamp": "2020-05-23T02:33:47.506Z"
            },
            "id": "urn:uuid:33045210-3ab8-4636-9b2a-c98d22ab915d",
            "label": "Sample 1",
            "remark": "Simple HTTP Service with Round-Robin Load Balancing",
            "schemaVersion": "3.0.0",
            "updateMode": "selective"
        },
        "changed": false,
        "failed": false,
        "queried": true
    }
}
```

Reply:
this should be under declaration, so as3_config['declaration'] this will allow us to do stuff like as3_config['tenants'] later on, without upsetting what the user is used to

Reply:
same applies to other endpoints: do_config['declaration'] ts_config['declaration'] etc this way you have wiggle room to add more stuff like 'info' (which i forsee as next step e.g. when user wants to check their ATC package versions etc)

Reply:
ok, I've added the declaration keyword as you've specified... Good point about future use cases...